File: functions/src/storageUrls.ts
import * as functions from 'firebase-functions/v1';
import * as admin from 'firebase-admin';
import {callable} from './callable';
import {
  assertStorageAccess,
  createPresignedUploadUrl,
  deleteObject,
  publicObjectUrl,
} from './spaces';

interface UploadUrlRequest {
  objectPath?: string;
  contentType?: string;
  contentLength?: number;
}

interface DeleteObjectRequest {
  objectPath?: string;
}

export const getStorageUploadUrl = callable().https.onCall(async (data, context) => {
  if (!context.auth) {
    throw new functions.https.HttpsError(
      'unauthenticated',
      'Sign in to upload files.',
    );
  }

  const payload = (data ?? {}) as UploadUrlRequest;
  const objectPath = String(payload.objectPath ?? '').trim();
  const contentType = String(payload.contentType ?? 'image/jpeg').trim();
  const contentLength = Number(payload.contentLength ?? 0);

  if (!objectPath) {
    throw new functions.https.HttpsError(
      'invalid-argument',
      'objectPath is required.',
    );
  }
  if (!contentType) {
    throw new functions.https.HttpsError(
      'invalid-argument',
      'contentType is required.',
    );
  }

  await assertStorageAccess(context.auth.uid, objectPath, contentType, contentLength);
  
  const spaces = await createPresignedUploadUrl({objectPath, contentType});
  
  const bucket = admin.storage().bucket();
  const file = bucket.file(objectPath);
  const [fallbackUploadUrl] = await file.getSignedUrl({
    action: 'write',
    expires: Date.now() + 15 * 60 * 1000,
    contentType,
  });
  const fallbackPublicUrl = `https://firebasestorage.googleapis.com/v0/b/${bucket.name}/o/${encodeURIComponent(objectPath)}?alt=media`;

  return {
    ...spaces,
    fallbackUploadUrl,
    fallbackPublicUrl,
  };
});

export const deleteStorageObject = callable().https.onCall(async (data, context) => {
  if (!context.auth) {
    throw new functions.https.HttpsError(
      'unauthenticated',
      'Sign in to delete files.',
    );
  }

  const payload = (data ?? {}) as DeleteObjectRequest;
  const objectPath = String(payload.objectPath ?? '').trim();
  if (!objectPath) {
    throw new functions.https.HttpsError(
      'invalid-argument',
      'objectPath is required.',
    );
  }

  await assertStorageAccess(context.auth.uid, objectPath);
  await deleteObject(objectPath);
  return {deleted: true, objectPath};
});

/** Converts a legacy Firebase Storage URL to the CDN URL when possible. */
export function firebaseUrlToCdnUrl(url: string): string | null {
  if (!url.includes('firebasestorage.googleapis.com')) return null;
  const marker = '/o/';
  const start = url.indexOf(marker);
  if (start === -1) return null;
  let encoded = url.substring(start + marker.length);
  const query = encoded.indexOf('?');
  if (query !== -1) encoded = encoded.substring(0, query);
  const objectPath = decodeURIComponent(encoded);
  if (!objectPath) return null;
  return publicObjectUrl(objectPath);
}

File: lib/core/services/storage_urls.dart
/// Helpers for DigitalOcean Spaces CDN and legacy Firebase Storage URLs.
class StorageUrls {
  StorageUrls._();

  static const cdnHost = 'talenthire-media.sgp1.cdn.digitaloceanspaces.com';
  static const originHost = 'talenthire-media.sgp1.digitaloceanspaces.com';

  static bool isStorageUrl(String url) {
    if (url.isEmpty) return false;
    return url.contains('firebasestorage.googleapis.com')
        || url.contains(cdnHost)
        || url.contains(originHost);
  }

  static String? objectPathFromUrl(String url) {
    if (url.isEmpty) return null;
    if (url.contains(cdnHost) || url.contains(originHost)) {
      try {
        final path = Uri.parse(url).path.replaceFirst('/', '');
        return path.isEmpty ? null : Uri.decodeComponent(path);
      } catch (_) {
        return null;
      }
    }
    if (!url.contains('firebasestorage.googleapis.com')) return null;
    const marker = '/o/';
    final start = url.indexOf(marker);
    if (start == -1) return null;
    var encoded = url.substring(start + marker.length);
    final query = encoded.indexOf('?');
    if (query != -1) encoded = encoded.substring(0, query);
    final decoded = Uri.decodeComponent(encoded);
    return decoded.isEmpty ? null : decoded;
  }
}
